src/P2P/MulticastBroadcaster.ts

```diff
@@ -0,0 +1,193 @@
+
+      const myPubKey = this.keyManager.getPublicKey()
+
+      if (decryptedRequestingPubKey.toString() == myPubKey) { // Response
+        // Make sure decrypted bytes equal raw bytes in memory
+        const originalMessage = this.peerPubKeyMessages.get(decryptedTargetPubKey.toString())?.rawRandomMessage
+
+        if (decryptedMessage.toString() == originalMessage?.toString()) {  // Validated!
+          // Add peer info to peerStore
+          const newPeerInfo = decodedMessage.responsePeerInfo
+          if (newPeerInfo) {
+            this.peerStore.add(newPeerInfo)
+            // Remove peerId from requested messages
+            const pubKey = newPeerInfo.publicKey
+            this.peerPubKeyMessages.delete(pubKey)
+            console.log(`New peer added to the store`);
+            this.emit('found', newPeerInfo)
+          } else {
+            this.emit('error', "I got a validated response. But no peerInfo");
+          }
+        }
+      } else {  // Requests on target node
+        // Try decrypting message
+        // Re-encrypt the data and send it on its way
+
+        const encryptedTargetPubKey = await this.keyManager.encryptData(Buffer.from(myPubKey), decryptedRequestingPubKey)
+        const encryptedMessage = await this.keyManager.encryptData(decryptedMessage, decryptedRequestingPubKey)
+        const encryptedPubKey = await this.keyManager.encryptData(decryptedRequestingPubKey, decryptedRequestingPubKey)
+        const handshakeMessage = RPCMessage.encodeHandshakeMessage(
+          Buffer.from(encryptedTargetPubKey),
+          Buffer.from(encryptedPubKey),
+          Buffer.from(encryptedMessage),
+          this.peerStore.localPeerInfo
+        )
+        this.socket.send(handshakeMessage, 0, handshakeMessage.length, UDP_MULTICAST_PORT, UDP_MULTICAST_ADDR);
+      }
+    } catch (err) { // Couldn't decode message
+      // We don't want the multicast discovery to error on every message it coudln't decode!
+    }
+  }
+
+  async requestPeerContact(pubKey: string) {
+    const pubKeyBuf = Buffer.from(pubKey)
+    const randomMessage = crypto.randomBytes(16)
+    // Encrypt message
+    const encryptedPeerPubKey = await this.keyManager.encryptData(pubKeyBuf, pubKeyBuf)
+    const encryptedRandomMessage = await this.keyManager.encryptData(randomMessage, pubKeyBuf)
+    const encryptedLocalPubKey = await this.keyManager.encryptData(Buffer.from(this.keyManager.getPublicKey()), pubKeyBuf)
+
+    // Add to peer messages to be sent over multicast
+    this.peerPubKeyMessages.set(pubKey, {
+      encryptedLocalPubKey: Buffer.from(encryptedLocalPubKey),
+      encryptedPeerPubKey: Buffer.from(encryptedPeerPubKey),
+      rawRandomMessage: randomMessage,
+      encryptedRandomMessage: Buffer.from(encryptedRandomMessage)
+    })
+  }
+
+  /**
+   * Stop sending queries to the LAN.
+   *
+   * @returns {Promise}
+   */
+  async stop(): Promise<void> {
```

Comment:
Perhaps we can use RAII for this? js-virtualfs had a similar issue here with cyclic dependencies, and thus had explicit constructors and deconstructors. But if we can model the construction/deconstruction of side-effects with the lifetime of objects, this keeps our implementation more cohesive with our programming style.

Reply:
I'm not sure how we would do that, perhaps we can discuss that at our next meeting?

Reply:
I have removed the stop function and moved the start functionality into the initialisation of the MulticastBroadcaster class.